### PR TITLE
RD2: Convert Closed RD having Date Established and Next Donation Date empty

### PR DESCRIPTION
**Pull request by RadSFDeveloper**
Convert closed legacy Recurring Donation having empty Date Established and Next Donation to the enhanced format.

# Critical Changes

# Changes

# Issues Closed

# Community Ideas Delivered

# New Metadata

# Deleted Metadata

### CODE REVIEW

**Comment by RadSFDeveloper** [comment]
@force, the PR #4969 has been merged so acceptance criteria involving Current and Next Year Value can be verified.

**Comment by RadSFDeveloper** [comment]
**lurch: attach W-039045

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x1E000001HrPZQA0" target="_blank">W-039045</a>

**Comment by RadSFDeveloper** [comment]
@force2b , the PR is ready for the code review.

File: src/classes/RD2_DataMigrationMapper.cls [hunk 1]
```diff
@@ -266,7 +266,7 @@ public class RD2_DataMigrationMapper {
             installmentDate = rd.npe03__Date_Established__c;
         }
 
-        return installmentDate.day().format();
+        return installmentDate == null ? null : installmentDate.day().format();
```
Comment (force2b) [inline src/classes/RD2_DataMigrationMapper.cls]: Is there other logic that will log an exception if this is an Active and Monthly RD and it's unable to determine a DayOfMonth?

Comment (RadSFDeveloper) [reply src/classes/RD2_DataMigrationMapper.cls]: To check if it is Active, Monthly and both Date Established and Next Donation Date are empty?

Comment (force2b) [reply src/classes/RD2_DataMigrationMapper.cls]: What I mean is if this is an Active RD and the InstallmentPeriod is Monthly we would normally require a value in DayOfMonth field. If this method now returns a null value to store into the DayOfMonth field what will happen during data migration?  Will the normal trigger validation throw an error that will get logged or would it just allow the change - which would end up leaving the data in invalid state?

Comment (RadSFDeveloper) [reply src/classes/RD2_DataMigrationMapper.cls]: If there is an Active RD, either Date Established or Next Donation Date need to be provided. If the "Always Use Last Day Of Month" is unchecked, "Day of Month" will be determined during migration correctly.

I added two new tests to verify "Effective Date", "Next Donation Date" and "Day of Month" when either date is missing, the RD has no Opportunities and the "Effective Date" is empty to start with. The "Effective Date" is determined on earliest Opportunity Close Date, or set to Date Established if no Opp exists.  See `shouldConvertOpenRDWhen*()` methods.

**Comment by RadSFDeveloper** [comment]
@Julian88Tex , the PR is ready for testing.

**Comment by Julian88Tex** [comment]
@RadSFDeveloper I've completed testing on this WI and will merge.

**Pull request merged**